pkg/sql/logictest/tests/5node/5node_test: TestLogic_distsql_agg failed

pkg/sql/logictest/tests/5node/5node_test.TestLogic_distsql_agg [failed](https://tanzanite.cluster.engflow.com/invocations/default/e85e45e0-d780-4e5e-81af-cc00311f0751?testReportRun=2&testReportShard=4&testReportAttempt=1#targets-Ly9wa2cvc3FsL2xvZ2ljdGVzdC90ZXN0cy81bm9kZTo1bm9kZV90ZXN0) on master @ [5d02bd9ff6b2bccecf6d43fc6cd647167b91f782](https://github.com/cockroachdb/cockroach/commits/5d02bd9ff6b2bccecf6d43fc6cd647167b91f782):

Fatal error:

```
panic: test timed out after 1h16m37s
running tests:
	TestLogic_distsql_agg (1h16m37s)

```
Stack: 

```
goroutine 1488270 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x265
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x45
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestLogic_distsql_agg
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestLogic_distsql_agg2932049972
    test_log_scope.go:81: use -show-logs to present logs inline
[10:42:53] setting distsql_workmem='24144B';
[10:42:57] --- progress: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_agg: 3 statements
[10:43:05] --- progress: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_agg: 4 statements
[10:43:12] --- progress: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_agg: 5 statements
[10:43:14] --- progress: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_agg: 6 statements
[10:43:16] --- progress: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_agg: 8 statements
[10:43:18] --- progress: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_agg: 10 statements
[10:43:21] --- progress: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_agg: 13 statements
[10:43:24] --- progress: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_agg: 16 statements
[10:43:27] --- progress: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_agg: 18 statements
[10:43:30] --- progress: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_agg: 20 statements
[10:43:33] --- progress: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_agg: 21 statements
[10:43:35] --- progress: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_agg: 22 statements
[10:43:38] --- progress: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_agg: 25 statements
[10:43:41] --- progress: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/5node/5node_test_/5node_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/distsql_agg: 29 statements
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=4</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_distsql_agg.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-38232

--- a/pkg/sql/colexec/colexecdisk/external_sort.go
+++ b/pkg/sql/colexec/colexecdisk/external_sort.go
@@ -259,15 +259,22 @@ func NewExternalSorter(
 	if testingVecFDsToAcquire > 0 {
 		maxNumberPartitions = testingVecFDsToAcquire
 	}
+	// Each disk queue will use up to BufferSizeBytes of RAM, so we'll reduce
+	// the memoryLimit of the partitions to sort in memory by those cache sizes.
+	diskQueueBuffersUsage := int64(maxNumberPartitions * diskQueueCfg.BufferSizeBytes)
 	if memoryLimit == 1 {
 		// If memory limit is 1, we're likely in a "force disk spill"
 		// scenario, but we don't want to artificially limit batches when we
 		// have already spilled, so we'll use a larger limit.
 		memoryLimit = execinfra.DefaultMemoryLimit
+	} else if budget := int64(100 << 10); memoryLimit < diskQueueBuffersUsage+budget {
+		// If the limit is not 1 but relatively small, we'll ensure that we have
+		// at least 100KiB of memory to work with. This memory is only utilized
+		// after we have spilled to disk, so we want to avoid sorting the data
+		// one tuple at a time (with each tuple forcing us to repartition).
+		memoryLimit = diskQueueBuffersUsage + budget
 	}
-	// Each disk queue will use up to BufferSizeBytes of RAM, so we reduce the
-	// memoryLimit of the partitions to sort in memory by those cache sizes.
-	memoryLimit -= int64(maxNumberPartitions * diskQueueCfg.BufferSizeBytes)
+	memoryLimit -= diskQueueBuffersUsage
 	// We give half of the available RAM to the in-memory sorter. Note that we
 	// will reuse that memory for each partition and will be holding onto it all
 	// the time, so we cannot "return" this usage after spilling each partition.
@@ -276,13 +283,6 @@ func NewExternalSorter(
 	inMemSortOutputLimit := inMemSortTotalMemoryLimit / 5
 	// We give another half of the available RAM to the merge operation.
 	mergeMemoryLimit := memoryLimit / 2
-	if inMemSortPartitionLimit < 1 {
-		// If the memory limit is 0, the input partitioning operator will return
-		// a zero-length batch, so make it at least 1.
-		inMemSortPartitionLimit = 1
-		inMemSortOutputLimit = 1
-		mergeMemoryLimit = 1
-	}
 	inputPartitioner := newInputPartitioningOperator(sortUnlimitedAllocator, input, inputTypes, inMemSortPartitionLimit)
 	var inMemSorter colexecop.ResettableOperator
 	if topK > 0 {

